doc/design/eventdriven-callbacks.md

```diff
@@ -0,0 +1,247 @@
+## Problem Description
+
+We require a public API so that a developer can register callbacks for certain
+events like the arrival of a sample.
+
+## Name of the design element
+
+ - ReactAl = React And Listen
+ - Reactor
+ - Eventler
+ - LiAR - Listen And React
+
+
+## Terminology
+
+ - **event driven** a one time reaction which is caused directly by an event.
+      Example: a new sample has been delivered to a subscriber.
+ - **state driven** a repeating reaction which is caused by a specific structural 
+      state until the state changes.
+      Example: a subscriber has stored samples which were not inspected by the user.
+ - **robust API** we use this term to describe an API which is hard 
+      to misuse. Especially important in concurrent code since a misuse can lead to race 
+      conditions, Heisenbugs and in general hard to debug problems.
+ - **robust API feature** a feature we only support to increase the robustness 
+      of the API and bugs caused by misuse.
+ - (Reactor pattern)[https://en.wikipedia.org/wiki/Reactor_pattern]
+ - (Condition Variable)[https://en.wikipedia.org/wiki/Monitor_(synchronization)#Condition_variables_2]
+ 
+## Design
+
+### General
+The usage should be similar to the _WaitSet_ with a key difference - it should 
+be **event driven** and not a mixture of event and state driven, depending on
+which event is attached, like in the _WaitSet_.
```

Comment:
I extended the first sentence in the document (Problem Description) for a better description.

At the moment we cannot link a WaitSet document since we do not have one and I would suggest to wait with the design documentation till this part is finalized since the WaitSet has to maybe adapt as well.